crates/circuit/src/dag_node.rs

```diff
@@ -76,114 +109,121 @@ pub struct DAGOpNode {
 impl DAGOpNode {
     #[new]
     #[pyo3(signature = (op, qargs=None, cargs=None, *, dag=None))]
-    fn new(
+    pub fn py_new(
         py: Python,
-        op: &Bound<PyAny>,
-        qargs: Option<&Bound<PySequence>>,
-        cargs: Option<&Bound<PySequence>>,
-        dag: Option<&Bound<PyAny>>,
-    ) -> PyResult<(Self, DAGNode)> {
-        let qargs =
-            qargs.map_or_else(|| Ok(PyTuple::empty_bound(py)), PySequenceMethods::to_tuple)?;
-        let cargs =
-            cargs.map_or_else(|| Ok(PyTuple::empty_bound(py)), PySequenceMethods::to_tuple)?;
-
-        let sort_key = match dag {
-            Some(dag) => {
-                let cache = dag
-                    .getattr(intern!(py, "_key_cache"))?
-                    .downcast_into_exact::<PyDict>()?;
-                let cache_key = PyTuple::new_bound(py, [&qargs, &cargs]);
-                match cache.get_item(&cache_key)? {
-                    Some(key) => key,
-                    None => {
-                        let indices: PyResult<Vec<_>> = qargs
-                            .iter()
-                            .chain(cargs.iter())
-                            .map(|bit| {
-                                dag.call_method1(intern!(py, "find_bit"), (bit,))?
-                                    .getattr(intern!(py, "index"))
-                            })
-                            .collect();
-                        let index_strs: Vec<_> =
-                            indices?.into_iter().map(|i| format!("{:04}", i)).collect();
-                        let key = PyString::new_bound(py, index_strs.join(",").as_str());
-                        cache.set_item(&cache_key, &key)?;
-                        key.into_any()
+        op: Bound<PyAny>,
+        qargs: Option<TupleLikeArg>,
+        cargs: Option<TupleLikeArg>,
+        #[allow(unused_variables)] dag: Option<Bound<PyAny>>,
```

Comment:
My only though is while this isn't used anymroe we may want to add it back so we can have a weakref to the parent dag that we can use for the setters to avoid the api impact around mutation by reference.

Reply:
I'm fine leaving it there if you think there's a chance we'll want it, though if we do backwards weakrefs, maybe we still wouldn't need it in the Python-space constructor, since it'd get attached only coming from Rust space?

Reply:
I guess it depends on the python space call. There are quite a few places we were previously doing something like:

```
new_node = DAGOpNode.from_instruction(
    CircuitInstruction.from_standard(op, ...),
    dag=new_dag
)
new_dag._apply_op_back(new_node)
```

that were updated to drop the `dag` arg. Although realistically we were only doing before for the faster sort key creation and caching which isn't relevant anymore. I think for 2.0 we'll probably want to drop this argument and `sort_key` as part of the public interface altogether.

Reply:
Sure, we were attaching it from Python space, but given the changes here, there was no longer a reason for us to need to.  In a world where `DAGOpNode` includes a weakref to the `DAGCircuit` that it came from, it feels like the only place that would need to set that weakref would be when the `DAGCircuit` object creates the `DAGOpNode`, since it'd also need to attach the node ID to make the backref actually work (i.e. so you can access the "true" weight).  So a `dag` argument on the Python-space constructor still would have no effect, because the node wouldn't have yet been attached to a dag, so the backref would be meaningless.

Reply:
Tracked in #13022.